src/sai/i2s.rs

```diff
@@ -0,0 +1,877 @@
+//! # Serial Audio Interface - Inter-IC Sound
+//!
+//! Inter-IC Sound.
+//!
+use core::convert::TryInto;
+
+use crate::rcc::{rec, CoreClocks, ResetEnable};
+use crate::sai::{GetClkSAI, Sai, SaiChannel, INTERFACE};
+use crate::stm32;
+use crate::stm32::{SAI1, SAI2, SAI3, SAI4};
+use crate::time::Hertz;
+
+use crate::device::sai4::ch::sr;
+use crate::gpio::gpioa::{PA0, PA1, PA12, PA2};
+use crate::gpio::gpiob::PB2;
+use crate::gpio::gpioc::{PC0, PC1};
+use crate::gpio::gpiod::{
+    PD0, PD1, PD10, PD11, PD12, PD13, PD14, PD15, PD4, PD6, PD8, PD9,
+};
+use crate::gpio::gpioe::{
+    PE0, PE11, PE12, PE13, PE14, PE2, PE3, PE4, PE5, PE6,
+};
+use crate::gpio::gpiof::{PF11, PF6, PF7, PF8, PF9};
+use crate::gpio::gpiog::{PG10, PG7, PG9};
+use crate::gpio::gpioh::{PH2, PH3};
+use crate::gpio::gpioi::{PI4, PI5, PI6, PI7};
+use crate::gpio::{Alternate, AF10, AF6, AF8};
+use stm32h7::Variant::Val;
+
+use embedded_hal::i2s::FullDuplex;
+
+const NUM_SLOTS: u8 = 16;
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum I2SMode {
+    Master = 0b00,
+    Slave = 0b10,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum I2SDir {
+    Tx = 0b00,
+    Rx = 0b01,
+}
+
+#[derive(Copy, Clone)]
+pub enum I2SDataSize {
+    BITS_8 = 0b001,
+    BITS_10 = 0b010,
+    BITS_16 = 0b100,
+    BITS_20 = 0b101,
+    BITS_24 = 0b110,
+    BITS_32 = 0b111,
+}
+
+#[derive(Copy, Clone)]
+enum I2SSlotSize {
+    BITS_16 = 0b01,
+    BITS_32 = 0b10,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum I2SProtocol {
+    MSB,
+    LSB,
+}
+
+#[derive(Copy, Clone)]
+pub enum I2SSync {
+    Master = 0b00,
+    Internal = 0b01,
+    External = 0b10,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum I2SError {
+    NoChannelAvailable,
+}
+
+pub enum I2SClockStrobe {
+    Rising,
+    Falling,
+}
+
+#[derive(Copy, Clone)]
+pub enum I2SCompanding {
+    Disabled = 0b00,
+    // Reserved = 0b01,
+    ALaw = 0b10,
+    MuLaw = 0b11,
+}
+
+pub enum I2SCompliment {
+    Ones = 0,
+    Twos = 1,
+}
+
+impl I2SCompliment {
+    fn as_bool(&self) -> bool {
```

Comment:
I would have expected to see `#[derive(Copy, Clone)]` on I2SComplement, and then to see this written as `impl From<I2SCompliment> for bool`

Not an essential change, would just make it more idiomatic